File: packages/repo-gc-typescript/src/heuristics/coupling.ts
import { Finding, FindingKind, Severity, type Thresholds } from 'repo-gc';
import { FileInfo, ImportGraph } from '../types';

export function analyze(
  info: FileInfo,
  graph: ImportGraph,
  threshold: Thresholds,
  idCounter: { value: number },
): Finding[] {
  const fanIn = graph.fanIn.get(info.path) || 0;
  const fanOut = graph.fanOut.get(info.path) || 0;

  if (fanIn < threshold.fanInLimit && fanOut < threshold.fanOutLimit) {
    return [];
  }

  const i = fanIn + fanOut > 0 ? fanOut / (fanIn + fanOut) : 0;
  const fanInOver = fanIn >= threshold.fanInLimit;
  const fanOutOver = fanOut >= threshold.fanOutLimit;

  let severity: Severity;
  let confidence: number;
  let patternLabel: string;

  if (fanInOver && !fanOutOver) {
    severity = Severity.Medium;
    confidence = 0.65;
    patternLabel = 'api';
  } else if (!fanInOver && fanOutOver) {
    severity = fanOut >= threshold.fanOutLimit * 3 ? Severity.High : Severity.Medium;
    confidence = 0.70;
    patternLabel = 'orch';
  } else {
    severity = fanIn >= threshold.fanInLimit * 3 || fanOut >= threshold.fanOutLimit * 2
      ? Severity.High : Severity.Medium;
    confidence = 0.85;
    patternLabel = 'god';
  }

  return [{
    id: `cp-${String(idCounter.value++).padStart(3, '0')}`,
    kind: FindingKind.CouplingHotspot,
    severity,
    confidence,
    path: info.relativePath,
    summary: '',
    reasons: [],
    evidence: [
      `fan_in: ${fanIn}`,
      `fan_out: ${fanOut}`,
      `instability: ${i.toFixed(3)}`,
      `pattern: ${patternLabel}`,
    ],
    suggested_next_step: '',
  }];
}

File: packages/repo-gc-python/python/repo_gc_python/heuristics/coupling.py
"""High fan-in/fan-out detection with pattern-aware output — mirrors Rust's coupling.rs."""

from ..cli import Threshold
from ..discovery import PythonFile
from ..graph import ImportGraph
from ..types import Finding, FindingKind, Severity


def analyze(
    pfile: PythonFile,
    graph: ImportGraph,
    threshold: Threshold,
    counter: int,
) -> Finding | None:
    fan_in = graph.fan_in.get(pfile.path, 0)
    fan_out = graph.fan_out.get(pfile.path, 0)
    in_limit = threshold.fan_in_limit
    out_limit = threshold.fan_out_limit

    if fan_in < in_limit and fan_out < out_limit:
        return None

    i = fan_out / (fan_in + fan_out) if (fan_in + fan_out) > 0 else 0.0
    fan_in_over = fan_in >= in_limit
    fan_out_over = fan_out >= out_limit

    if fan_in_over and not fan_out_over:
        severity = Severity.Medium
        confidence = 0.65
        pattern_label = "api"
    elif not fan_in_over and fan_out_over:
        if fan_out >= out_limit * 3:
            severity = Severity.High
        else:
            severity = Severity.Medium
        confidence = 0.70
        pattern_label = "orch"
    else:
        if fan_in >= in_limit * 3 or fan_out >= out_limit * 2:
            severity = Severity.High
        else:
            severity = Severity.Medium
        confidence = 0.85
        pattern_label = "god"

    return Finding(
        id=f"ch-{counter:03d}",
        kind=FindingKind.CouplingHotspot,
        severity=severity,
        confidence=confidence,
        path=str(pfile.relative_path),
        summary="",
        reasons=[],
        evidence=[
            f"fan_in: {fan_in}",
            f"fan_out: {fan_out}",
            f"instability: {i:.3f}",
            f"pattern: {pattern_label}",
        ],
        suggested_next_step="",
        estimated_tokens=None,
    )
